cli/flags.go

```diff
@@ -142,14 +148,44 @@ func (f *FlagSet) NewSection(name string) *FlagSection {
 	return fs
 }
 
+// AfterParse defines a post-parse function. This can be used to set flag
+// defaults that should not be set until after parsing, such as defaulting flag
+// values to the value of other flags. These functions are called after flags
+// have been parsed by the flag library, but before [Parse] returns.
+func (f *FlagSet) AfterParse(fn AfterParseFunc) {
+	if fn != nil {
+		return
+	}
+
+	f.afterParseFuncs = append(f.afterParseFuncs, fn)
+}
+
+// Arg implements flag.FlagSet#Arg.
+func (f *FlagSet) Arg(i int) string {
+	return f.flagSet.Arg(i)
+}
+
 // Args implements flag.FlagSet#Args.
 func (f *FlagSet) Args() []string {
 	return f.flagSet.Args()
 }
 
+// Lookup implements flag.FlagSet#Lookup.
+func (f *FlagSet) Lookup(name string) *flag.Flag {
+	return f.flagSet.Lookup(name)
+}
+
 // Args implements flag.FlagSet#Parse.
 func (f *FlagSet) Parse(args []string) error {
-	return f.flagSet.Parse(args)
+	// Call the normal parse function first, so that Args and everything are
+	// properly set for any after functions.
+	merr := f.flagSet.Parse(args)
```

Comment:
I wonder if we should return err right away if `Parse` fails. `afterParseFuncs` likely depend on the fully parsed flags. Otherwise, the error from the `afterParseFuncs` might not even make sense.

Reply:
Hmm - this is a good point. However, some developers might want to return "all" CLI errors at once instead of forcing the user to go through a bunch of validations one-by-one.

I suppose the signature for afterFunc could include `parseErr err`, but it feels a bit weird.

Reply:
Since we see use cases for both, I guess this question remains: is this functionality broadly useful, or would it make more sense to provide examples of how developers can do this themselves with a private function?

Reply:
Good question! I often find myself defining a struct to hold all flags (Dave also recent did this in abc CLI). I'm thinking an ideal experience like:

```go
set := cli.NewFlagSet()

type MyFlags struct {
  // The fields will be my flags
}

// Interface - minimally needed to register as a flag section.
func (f *MyFlags) Register(s *cli.FlagSet) {
  // Create a new section, register flags.
}

// This implements post parse hook interface.
func (f *MyFlags) AfterParse() error {
  // Post parse hook. Defaulting or validation.
}

f1 := &MyFlags{}

// Register a new section
// This calls the Register() above.
set.Register(f1)

// The parse calls AfterParse()
err := set.Parse(args)
```

Reply:
Hmm - I don't love magical interfaces as an API construct. I think I prefer [how Guardian does it](https://github.com/abcxyz/guardian/blob/main/pkg/flags/github.go#L28) where each flagset is manually registered in the call to `Flags()`. Because of how flags get added, we also don't have access to `MyFlags` during parsing, so it'd need to be something like:

```go
type MyFlags struct {
  flagA string
  flagB string
}

func (f *MyFlags) Register(set *cli.FlagSet) {
  f := set.NewSection("MY FLAGS")

  f.StringVar(...)

  set.AfterFunc(func(args []string) error {
    if f.flagA == "" {
      return fmt.Errorf("-flag-a is required")
    }
  })
}
```

Reply:
> "I don't love magical interfaces as an API construct"

I meant that I prefer to avoid situations in which a specially-named function changes the behavior of an upstream call. It's fine to implement an interface and then accept that type as an interface:

```go
type Fooer interface {
  Foo()
}

func DoSomething(f Fooer) {
  f.Foo()
}
```

but I dislike (and think it's a bad experience) when functions do something magic with interfaces (example):

```go
type Fooer interface {
  Foo()
}

func DoSomething(f Fooer) {
  if t, ok := f.(interface{
    Bar()
  }); ok {
    t.Bar()
  } else {
    t.Foo()
  }
}
```

because then suddenly defining a function on your struct can cause an unintended behavior change.

> That's also how JVS does it these days...

I think we've subtly converged on a similar pattern. However, I think there's some CUJs that remain unsolved in that pattern:

- If I have a separate flags struct, how and where do I provide validation for that structure? Should the validation live alongside my flagset or should it be in the parent command?

- If I need to do other post-processing (such as deferred defaulting), where does that happen? What if it needs to be reusable?

Reply:
> If I have a separate flags struct, how and where do I provide validation for that structure? Should the validation live alongside my flagset or should it be in the parent command?

I tend to define the validation alongside my flags struct. But there could be use cases where defining it in the command makes more sense. I think the approach in the PR provides the flexibility to do either.

> If I need to do other post-processing (such as deferred defaulting), where does that happen? What if it needs to be reusable?

With `AfterFunc`, I would default to have defaulting and validation logic defined as `AfterFunc` but will keep them alongside my flag struct. Like your example here: https://github.com/abcxyz/pkg/pull/168#discussion_r1274107967

That said, should we merge this PR?

---

Thanks for the explanation. I actually regard the interface magic as a Go strength :) Not your exact example, but something like:

```go
func DoSomething(f Fooer) {
  if t, ok := f.(interface{
    SetDefault()
  }); ok {
    // If the input has no defaulting logic,
    // then it doesn't have to implement an empty SetDefault func.
    t.SetDefault()
  }
  t.Foo()
}
```

Reply:
> That said, should we merge this PR?

I think there's still the remaining question about whether we run `AfterFunc`s when normal flag parsing fails.

Reply:
> Since we see use cases for both, I guess this question remains: is this functionality broadly useful, or would it make more sense to provide examples of how developers can do this themselves with a private function?

I think it's useful. If I had this, I would have put the defaulting/validation logic there. Like your example:

```go
func (f *MyFlags) Register(set *cli.FlagSet) {
  f := set.NewSection("MY FLAGS")

  f.StringVar(...)

  set.AfterFunc(func(args []string) error {
    if f.flagA == "" {
      return fmt.Errorf("-flag-a is required")
    }
  })
}
```

Reply:
Okay - I updated the code and the examples, plus added tests.